src/store.ts

```diff
@@ -0,0 +1,921 @@
+          req.id,
+          req.id,
+          req.recv,
+          req.timeout,
+        );
+
+        return {
+          kind: "createSubscription",
+          promise: promise,
+          callback: callback,
+        };
+      }
+      case "createCallback": {
+        const { promise, callback } = this.promises.callback(
+          req.id,
+          req.rootPromiseId,
+          req.recv,
+          req.timeout,
+        );
+
+        return { kind: "createCallback", promise: promise, callback: callback };
+      }
+
+      case "claimTask": {
+        const { task, applied } = this.tasks.transition(
+          req.id,
+          "claimed",
+          undefined,
+          undefined,
+          undefined,
+          undefined,
+          req.counter,
+          req.processId,
+          req.ttl,
+        );
+
+        util.assert(applied);
+
+        switch (task.type) {
+          case "invoke": {
+            const promise = this.promises.get(task.rootPromiseId);
+            return {
+              kind: "claimedtask",
+              message: {
+                kind: task.type,
+                promises: {
+                  root: { id: promise.id, data: promise },
+                },
+              },
+            };
+          }
+          case "resume": {
+            const rootPromise = this.promises.get(task.rootPromiseId);
+            const learPromise = this.promises.get(task.leafPromiseId);
+
+            return {
+              kind: "claimedtask",
+              message: {
+                kind: task.type,
+                promises: {
+                  root: { id: rootPromise.id, data: rootPromise },
+                  leaf: { id: learPromise.id, data: learPromise },
+                },
+              },
+            };
+          }
+        }
+      }
+      case "completeTask": {
+        const { task, applied } = this.tasks.transition(
+          req.id,
+          "completed",
+          undefined,
+          undefined,
+          undefined,
+          undefined,
+          req.counter,
+        );
+
+        util.assert(applied);
+
+        return {
+          kind: "completedtask",
+          task: {
+            id: task.id,
+            counter: task.counter,
+            timeout: 0,
+            processId: task.pid,
+            createdOn: task.createdOn,
+            completedOn: task.completedOn,
+          },
+        };
+      }
+      case "heartbeatTasks": {
+        let applied = false;
+        let affectedTasks = 0;
+
+        for (const task of this.tasks.scan()) {
+          if (task.state !== "claimed" || task.pid !== req.processId) {
+            continue;
+          }
+
+          applied = this.tasks.transition(
+            task.id,
+            "claimed",
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            true,
+          ).applied;
```

Comment:
Seems like `tasks.transition` will be better if we make it take and object with optional fields, that way it is not necessary to pass `undefined` to all the optional arguments, would look something like
```suggestion
          let {applied} = this.tasks.transition({
            id: task.id,
            state: "claimed",
            force: true,
          });
```